lib/instream/connection.ex

```diff
@@ -184,4 +194,11 @@ defmodule Instream.Connection do
   Usable options depend on the writer module configured.
   """
   @callback write(payload :: Line.point() | [Line.point()], opts :: Keyword.t()) :: any
+
+  @doc """
+  Deletes data from an InfluxDB bucket.
+
+  Usable options depend on the delete module configured.
+  """
+  @callback delete(payload :: map(), opts :: Keyword.t()) :: any
```

Comment:
Can we somehow tighten what is allow to pass as a `payload` here?

Right now I am thinking of a simple custom type like this:

```elixir
  @type delete_request :: %{
          required(:start) => binary,
          required(:stop) => binary,
          optional(:predicate) => binary
        }
```

This would allow the generated documentation to contain information about what the `payload` should be. Otherwise one would need to dig through the code to find out how to use it.

Reply:
I can create encoder `lib/instream/encoder/predicate.ex` that will have this type.
And it will just pattern match required and optional field and do Jason encode.
```elixir
defmodule Instream.Encoder.Predicate do
  @moduledoc """
  Encoder for the InfluxDB predicate protocol.
  """

  @type points ::
          %{
            required(:start) => binary,
            required(:stop) => binary,
            optional(:predicate) => map
          }

  @doc """
  Creates protocol contents for delete data points.
  """
  @spec encode(points()) :: binary
  def encode(points), do: encode_points(points)

  defp encode_points(%{"start" => start, "stop" => stop, "predicate" => predicate}) do
    Jason.encode!(%{start: start, stop: stop, predicate: predicate})
  end

  defp encode_points(%{"start" => start, "stop" => stop}) do
    Jason.encode!(%{start: start, stop: stop})
  end
end
```
Is something like this ok?
And if it is, do we want it to have catch all `encode_points` that return something or just let it fail?

Reply:
_First I would drop those binary keys completely, they do not make any sense if everything else is using atom keys._

The requirements for the map could be checked by the `Instream.Connection.delete/2` function defined in the `__using__` macro without any extra modules necessary:

```
def delete(%{start: _, stop: _} = payload, opts)
```

The function specification already tells the user what to pass so I do not think it hurts to enforce the requirements as early as possible. Any extra map keys that should be ignored can be dropped using `Map.take(payload, [:start, :stop, :predicate])`.

Reply:
Thank you so much. Fixed.